src/lib.rs

```diff
@@ -274,9 +275,92 @@ fn parse_unary_operator(input: ParseInput) -> ParseResult<UnaryOperator> {
+    ))
+    .parse(input)
+}
+
+/// Parses a binary operator and the expression that follows it, given the expression preceding
+/// the operator, returning all of this is a binary `Expression`.
+///
+/// E.g. left=`u1:1`, input=`&& u1:1`, returns the Expression for `u1:1 && u1:1`
+///
+/// parse_infix_expression handles any and all kinds of expressions that would
+/// left-recursively call parse_expression
+fn parse_infix_expression(input: ParseInput, left: Expression) -> ParseResult<Expression> {
+    // this implementation follows the 'Top Down Operator Precedence' algorithm. See
+    // <https://btmc.substack.com/p/how-to-parse-expressions-easy> or <https://tdop.github.io/>
+    //
+    // If we had any right-associative operators, we would have to use the 'precedence
+    // lowering' trick found here: https://btmc.substack.com/p/how-to-parse-expressions-easy
+    // Like so: right = parse_expression(tokens, get_precedence(op.tag)-1)
+
+    // TODO I'm going to try and use more combinators/less explicit.
+    // TODO use fn flat_map<G, H, O2>(self, g: G) -> FlatMap<Self, G, O>
+    let (rest, start) = position_ws().parse(input)?;
+    let (rest, op) = parse_binary_operator(rest)?;
+    let (rest, right) = parse_expression(rest, Some(op.thing))?;
+    let (_, end) = nom_locate::position(rest)?;
+    let re = RawExpression::from((left, op, right));
+    Ok((
+        rest,
+        Spanned {
+            span: Span::new(start, end),
+            thing: re,
+        },
+    ))
+}
+
+/// Parses an expression (e.g. binary, unary, arbitrarily nested expressions), given the
+/// preceding binary operator (if one exists), returning all of this is a binary `Expression`.
+///
+/// E.g. input=`u1:1 && u1:1`, previous=`Some(||)`, will return the `Expression` `u1:1 && u1:1`
+/// because `&&` has higher precedence than `||`.
+fn parse_expression(
+    input: ParseInput,
+    previous: Option<RawBinaryOperator>,
+) -> ParseResult<Expression> {
+    // this implementation follows the 'Top Down Operator Precedence' algorithm. See
+    // <https://btmc.substack.com/p/how-to-parse-expressions-easy> or <https://tdop.github.io/>
+
+    let higher_precedence_than_previous = |current: BinaryOperator| -> bool {
+        match (previous, current.thing) {
+            // when no previous, then we judge current is higher precedence
+            (None, _) => true,
+            (Some(previous), current) => match current.partial_cmp(&previous) {
+                Some(Greater) => true,
+                // no comparison, equal, and less are all judged not higher precedence
+                _ => false,
+            },
+        }
+    };
+
+    // Note: I could not figure out how to do the following using only combinators
+    let (mut rest, mut left) = parse_unary_atomic_expression(input)?;
+    let mut op: Option<BinaryOperator>;
+    loop {
+        (rest, op) = peek(opt(parse_binary_operator))(rest)?;
+        match op {
+            Some(op) => {
```

Comment:
Will this work for, e.g., match?
```
let foo = a + match(bleh) { ... }
```

I don't think it slots in as it stands, but maybe parse_infix_expression just needs to be up-beefed. Just wanted to get your thoughts on that.

Reply:
If I understand your question correctly:
Yes, Rust `match` is an expression so you can use it in a subexpression.

At the moment I'm not sure how to restructure parse_expression to make it more combinator style. I'll think about it. If you have any more suggestions, I'm happy to hear them, brainstorm with you, etc.